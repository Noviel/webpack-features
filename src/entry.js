const hotReloadEntry = 'webpack-hot-middleware/client';
const reactHotEntry = 'react-hot-loader/patch';
const polyfillEntry = 'babel-polyfill';

export default (
  { target, production },
  entries,
  {
    polyfill = target.browsers === 'legacy',
    hot = !!(!production && target.browsers),
    react = true,
  } = {}
) => {
  const result = {};

  Object.keys(entries).forEach(v => {
    result[v] = []
      .concat(polyfill ? polyfillEntry : [])
      .concat(hot && react ? reactHotEntry : [])
      .concat(hot ? hotReloadEntry : [])
      .concat(entries[v]);
  });

  return result;
};
