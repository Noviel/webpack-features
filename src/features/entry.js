const builtinEntries = {
  hot: {
    express: 'webpack-hot-middleware/client',
    react: 'react-hot-loader/patch',
  },
  polyfill: '@babel/polyfill',
};

export default (
  { target, production },
  {
    entries,
    polyfill = target.name === 'browsers' && target.value === 'legacy',
    hot = !production && target.name === 'browsers',
    react = true,
    express = false,
  } = {}
) => {
  if (Array.isArray(entries) || typeof entries === 'string') {
    entries = {
      index: entries,
    };
  } else if (!entries || typeof entries !== 'object') {
    throw new Error(`'entries' should be an object, an array or a string`);
  }

  return {
    entry: Object.keys(entries).reduce((acc, curr) => {
      acc[curr] = []
        .concat(polyfill ? builtinEntries.polyfill : [])
        .concat(hot && react ? builtinEntries.hot.react : [])
        .concat(hot && express ? builtinEntries.hot.express : [])
        .concat(entries[curr]);

      return acc;
    }, {}),
  };
};
