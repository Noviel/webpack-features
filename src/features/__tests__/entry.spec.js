import entry from '../entry';

const envs = {
  modernBrowsersDev: {
    target: { browsers: 'modern' },
    production: false,
  },

  modernBrowsersProd: {
    target: { browsers: 'modern' },
    production: true,
  },

  legacyBrowsersDev: {
    target: { browsers: 'legacy' },
    production: false,
  },

  legacyBrowsersProd: {
    target: { browsers: 'legacy' },
    production: true,
  },

  nodeProd: {
    target: { node: 'current' },
    production: true,
  },
};

describe('entry', () => {
  describe('should produce correct entries with default options', () => {
    it('for modern browsers in production mode', () => {
      expect(
        entry(envs.modernBrowsersProd, { index: 'index.js' })
      ).toMatchSnapshot();
    });
    it('for modern browsers in development mode', () => {
      expect(
        entry(envs.modernBrowsersDev, { index: 'index.js' })
      ).toMatchSnapshot();
    });
    it('for legacy browsers in production mode', () => {
      expect(
        entry(envs.legacyBrowsersProd, { index: 'index.js' })
      ).toMatchSnapshot();
    });
    it('for legacy browsers in development mode', () => {
      expect(
        entry(envs.legacyBrowsersDev, { index: 'index.js' })
      ).toMatchSnapshot();
    });
  });
});
