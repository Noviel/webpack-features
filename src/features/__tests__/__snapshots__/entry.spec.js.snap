// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entry should produce correct entries with default options for legacy browsers in development mode 1`] = `
Object {
  "entry": Object {
    "index": Array [
      "babel-polyfill",
      "react-hot-loader/patch",
      "webpack-hot-middleware/client",
      "index.js",
    ],
  },
}
`;

exports[`entry should produce correct entries with default options for legacy browsers in production mode 1`] = `
Object {
  "entry": Object {
    "index": Array [
      "babel-polyfill",
      "index.js",
    ],
  },
}
`;

exports[`entry should produce correct entries with default options for modern browsers in development mode 1`] = `
Object {
  "entry": Object {
    "index": Array [
      "react-hot-loader/patch",
      "webpack-hot-middleware/client",
      "index.js",
    ],
  },
}
`;

exports[`entry should produce correct entries with default options for modern browsers in production mode 1`] = `
Object {
  "entry": Object {
    "index": Array [
      "index.js",
    ],
  },
}
`;
